doc/developer/design/20230721_persist_fast_path.md

```diff
@@ -0,0 +1,83 @@
+
+## Goals
+
+- Fast responses to `SELECT * FROM large_collection LIMIT small_n` queries.
+  -  In particular, we want the cost of this query to be proportional to `small_n` and not the size of `large_collection`. (The constant factors may be much higher than the equivalent query against an index, though.)
+
+## Non-Goals
+
+- Improving the performance of `LIMIT` clause on a more general range of queries.
+
+## Overview
+
+Supporting a fast query path for select-limit queries against Persist shards will require changes to both our query planning and execution.
+
+- In `environmentd`, generalize the existing fast-path planning logic. As of today, fast-path plans were only generated when the queried collection has at least one index. We’ll extend this logic to also generate these plans for unindexed collections that are backed by a Persist shard.
+- When `clusterd` receives a peek against a Persist-backed collection — as opposed to an arrangement, which is always the case today — it will create a new `PersistPeek` operation that streams the data from a Persist shard. While it’s possible to efficiently and incrementally stream consolidated data from a Persist shard, no such API exists today; we’ll need to add one.
+
+## Detailed description
+
+### Planning
+
+Today, `environmentd` will generate a fast-path plan when a dataflow can be represented as: a `Get` from an indexed collection, a map-filter-project stage, and some “finishing” logic including order-by/limit. We’ll change this to also allow the `Get` to get from a non-indexed collection backed by a Persist shard.
+
+If we want to ensure that the cost of the query is proportional to the result set size and not the size of the collection, we'll need to introduce some additional restrictions:
+- No filters. Index-based fast-path queries are often used for single-key lookups, like `select * from indexed_collection where id = 'XXX'`. Persist cannot currently support efficient queries with this shape.
+- No `ORDER BY`. This would require us to load and sort the full collection, then apply the limit.
+
+We may be able to relax these requirements in the future. (For example, if we order the Persist data by the data's "logical" ordering and not its serialized representation, we could support primary-key lookups more efficiently.)
+
+Today, every select query generates a peek request into an arrangement. (Even non-fast-path queries work by creating a new dataflow that writes out its data into an arrangement, then peeking into that arrangement.) This design introduces a second type of peek: one that peeks directly into the backing Persist shard. Concrentely, this means that `environmentd` will send `ComputeCommand::Peek` commands that reference a persisted collection’s `Id`, not the id of an index or arrangement.
```

Comment:
It seems like a new `ComputeCommand` variant would be preferable over re-using `Peek`, considering that persist peeks have different constraints (e.g. no filters), different target specification (a persist location rather than an index ID), and vastly different handling on the cluster size.